Sources/CvPlayerAI.cpp

```diff
@@ -14395,15 +14498,15 @@ int CvPlayerAI::AI_unitValue(UnitTypes eUnit, UnitAITypes eUnitAI, CvArea* pArea
 			break;
 		}
 	}
-	
+
 	if ((iCombatValue > 0) && ((eUnitAI == UNITAI_ATTACK) || (eUnitAI == UNITAI_ATTACK_CITY)))
 	{
 		if (pArea != NULL)
 		{
 			AreaAITypes eAreaAI = pArea->getAreaAIType(getTeam());
 			if (eAreaAI == AREAAI_ASSAULT || eAreaAI == AREAAI_ASSAULT_MASSING)
 			{
-				for (int iI = 0; iI < GC.getNumPromotionInfos(); iI++)
+				for (iI = 0; iI < GC.getNumPromotionInfos(); iI++)
```

Comment:
This is wrong, variable should always be defined it the scope where it is used. For loop index this means in the loop variable as it was originally.

Reply:
Please clarify as there has been some differences of opinion among various programmers on this matter:
is (int iI = 0... or (iI = 0.... (with the int iI being declared at the beginning of the function instead of within the for statement) correct.  I'm not sure what you mean by defined in the scope where it is used.  My understanding was that it should be (int iI = 0...  But this is very much not the standard that Firaxis started with, which was to declare iI at the beginning of the function usually.  I don't usually fight that if that's how the function is when I write a For loop.

Is there a reason for one to be considered better form than the other and what are the advantages/disadvantages?  (can't promise I'll understand the answer but I'll try.)

Reply:
> Please clarify as there has been some differences of opinion among various programmers on this matter:
Well you know what they say about opinions! If you want to know the best practice on something you should trust google/stackoverflow/expert blog posts (there is a large array of experts in C++ who write on these kind of things, e.g. Scott Meyers), over the opinions of any individual programmer.
When I talk about these things related to style, structure etc., I'm relaying the information I learned from these sources, and confirmed in my own experience as being good. i.e. You can confirm the things I'm saying with google, and where you can't I will be happy to update my understanding of best practices. 

> is (int iI = 0... or (iI = 0.... (with the int iI being declared at the beginning of the function instead of within the for statement) correct. 
The first is always correct. You should never be defining variables at the start of a function, this is a holdover from very old C style programming when it was required to declare variables at the start of a scope (C89 required this, but not C99, that is how old we are talking). Lots of C++ programmers in the early days were C programmers who didn't bother to learn how to write C++ correctly, because they thought it was basically the same thing but with classes, or they preferred to stick with the style they knew.

> I'm not sure what you mean by defined in the scope where it is used.
A scope in C++ is basically everything `{` between two matching curly braces `}`. Other structures can define a scope, but that is the main one in functions. So in a for loop the scope of the for loop is the `{ ... }` that is executed for each iteration.
I guess I actually meant to say "declared in the scope where it is used.", rather than defined. In other words if you are using a variable inside a for loop, and it's only used there, you should declare it there as well.
i.e. Instead of:
```cpp
int i, value;
for(i = 0; i < 10; i++)
{
    value = 0;
    value += get_value(i);
    if(value > 5) 
    {
        max_value = value;
        // etc
    }
}
```
You should write:
```cpp
for(int i = 0; i < 10; i++)
{
    int value = get_value(i);
    if(value > 5) 
    {
        max_value = value;
        // etc
    }
}
```

> Is there a reason for one to be considered better form than the other and what are the advantages/disadvantages?

As I alluded to above, I'm not giving my random opinions anywhere here when it comes to what is good coding practice or style (when I do I will be clear that it is just my opinion rather than established best practices), these are conventions of good software engineering, and are well supported by reasoning and experience of the community as a whole.

One reason I can give, that is particularly salient in our case, is that I already found a few bugs in the code that were caused due to declaring variables at the start of a function instead of where they are used. For instance in one case a variable was used in a for loop (and of course declared at the start of the function), but the coder forgot to reset the value at the start of each loop, so it ended up corrupting the final calculation. If they had declared the variable at the start of the loop then the compiler would have warned them if they didn't initialize it before using it. In other cases the same variable gets reused in different places, but the person forgot to reset it, so the value in the second location is corrupted by the value in the first. In this case it is *two* bad coding practices, firstly declaring the variable before it is used, and secondly reusing a variable for two different purposes.

Another reason is readability. Readers expect variables to be declared in the scope in which they are used, so seeing a variable being declared indicates that it is used in that scope. It is much easier to understand the use of a variable when it is declared where it is used. i.e. You know it is only used in that scope. Another way to phrase this (which applies to a lot of style guidelines), is that code should clearly (and unambiguously, and concisely) indicate the intent of the programmer. 

Performance is another reason: declaring a variable is allocating it on the stack. That is fast, but it isn't free. If you declare a bunch of variables and then check an if statement and leave the function before using them, you wasted time allocating them.

And again, it confuses the optimizer. It wants to optimize where variables are allocated, what registers they are assigned to etc., and this is much easier for it when it can easily determine what section of the code the variable is actually used in. Declaring all the variables at the top of the function is you telling the optimizer that you could be using these variables anywhere.

Reply:
> But this is very much not the standard that Firaxis started with, which was to declare iI at the beginning of the function usually. I don't usually fight that if that's how the function is when I write a For loop.

I mentioned already I guess, but I would not use anything in this code base as an example of how to write good code. Its not even a matter of just using the existing code style in the case of all other things being equal. In this case continuing to use the existing code style is actually bad for the health of the code and the future maintainability. It's better to start any new code or functions using correct style and practices, and to update existing functions as and when we can. e.g. I will update any function I am significantly modifying. I already fixed large portions of CvCity and CvUnit classes (not all though).

Also a further note on style and correctness in C++: there are many professional high quality code bases available for browsing on github. You can delve into any of them to see various examples of what good code should look like. e.g. I tried just now scrolling through a few files in the Unreal code base (which has some problems, but is pretty good generally) to look for if nesting. I couldn't find any more than 2 deep, and the ones that *were* 2 deep without else statements were done for optimization purposes, e.g. where the result from the inner if statement is used in the the inner block, like:
```cpp
if(a > b) 
{
  Ptr* obj = get_best_ptr(a, b);
  if(obj != nullptr)
  {
    obj->do_a_thing();
  }
}
```
However even in this case it is questionable. Perhaps `get_best_ptr` and the following if statement should instead by replaced by `try_do_a_thing_to_base_ptr(a, b);`.